sound/soc/sof/virtio-miscdev.c

```diff
@@ -0,0 +1,255 @@
+
+	list_del(&vbe->list);
+	devm_kfree(vbe->sdev->dev, vbe);
+	f->private_data = NULL;
+	return 0;
+}
+
+int snd_sof_virtio_miscdev_register(struct snd_sof_dev *sdev)
+{
+	struct virtio_miscdev *vaudio;
+	int ret;
+
+	ret = snd_audio_virtio_miscdev_register(sdev->dev, sdev, &vaudio);
+	if (ret)
+		return ret;
+
+	vaudio->open = sof_virtio_open;
+	vaudio->ioctl = sof_virtio_ioctl;
+	vaudio->release = sof_virtio_release;
+
+	return 0;
+}
+
+int snd_sof_virtio_miscdev_unregister(void)
+{
+	return snd_audio_virtio_miscdev_unregister();
+}
+
+static int vbs_audio_open(struct inode *inode, struct file *f)
+{
+	struct virtio_miscdev *vaudio = get_virtio_audio();
+
+	if (!vaudio)
+		return -ENODEV;	/* This should never happen */
+
+	dev_dbg(vaudio->dev, "virtio audio open\n");
+	if (vaudio->open)
+		return vaudio->open(f, virtio_audio->priv);
+
+	return 0;
+}
+
+static long vbs_audio_ioctl(struct file *f, unsigned int ioctl,
+			    unsigned long arg)
+{
+	struct virtio_miscdev *vaudio = get_virtio_audio();
+
+	if (!vaudio)
+		return -ENODEV;	/* This should never happen */
+
+	dev_dbg(vaudio->dev, "virtio audio ioctl\n");
+	if (vaudio->ioctl)
+		return vaudio->ioctl(f, vaudio->priv, ioctl, arg);
+	else
+		return -ENXIO;
+}
+
+static int vbs_audio_release(struct inode *inode, struct file *f)
+{
+	struct virtio_miscdev *vaudio = get_virtio_audio();
+
+	if (!vaudio)
+		return -ENODEV;	/* This should never happen */
+
+	dev_dbg(vaudio->dev, "release virtio audio\n");
+
+	if (vaudio->release)
+		vaudio->release(f, vaudio->priv);
+
+	return 0;
+}
+
+static const struct file_operations vbs_audio_fops = {
+	.owner          = THIS_MODULE,
+	.release        = vbs_audio_release,
+	.unlocked_ioctl = vbs_audio_ioctl,
+	.open           = vbs_audio_open,
+	.llseek		= noop_llseek,
+};
+
+static struct miscdevice vbs_audio_k = {
+	.minor = MISC_DYNAMIC_MINOR,
+	.name = "vbs_k_audio",
+	.fops = &vbs_audio_fops,
+};
+
+/**
+ * snd_audio_virtio_miscdev_register() - init the virtio be audio driver
+ * @dev: the device the audio virtio belongs to
+ * @data: the priv data of virtio_miscdev
+ * @va: allocated virtio_miscdev, return to caller
+ *
+ * This function registers the misc device, which will be used
+ * by the user space to communicate with the audio driver.
+ *
+ * Return: 0 for success or negative value for err
+ */
+int snd_audio_virtio_miscdev_register(struct device *dev, void *data,
+				      struct virtio_miscdev **va)
+{
+	struct virtio_miscdev *vaudio;
+	int ret;
+
+	ret = misc_register(&vbs_audio_k);
+	if (ret) {
+		dev_err(dev, "misc device register failed %d\n", ret);
+		return ret;
+	}
+
+	vaudio = kzalloc(sizeof(*vaudio), GFP_KERNEL);
+	if (!vaudio) {
+		misc_deregister(&vbs_audio_k);
+		return -ENOMEM;
+	}
+
+	vaudio->priv = data;
+	vaudio->dev = dev;
+	virtio_audio = vaudio;
+	*va = vaudio;
+
+	// TODO: must be a way to regisuer misc dev with private data
+	//  so global not needed
+	return 0;
+}
+EXPORT_SYMBOL(snd_audio_virtio_miscdev_register);
```

Comment:
Does this need to be exported, it looks like it's only called internally ? (and should be static)

Reply:
This is defined by Libin, and he said he will send BE patches himself, so I just remove it in the patch